https-bastion.md

```diff
@@ -0,0 +1,137 @@
+# HTTPS Bastion
+
+This document specifies an HTTPS bastion, a service that acts as a reverse proxy
+accepting connections from both clients and backends and routing requests from
+the former to the latter.
+
+Bastions allow backends to run in an environment without a publicly reachable
+address, while still being reachable synchronously over a public HTTP endpoint.
+
+The design requires no changes to HTTPS clients, and can be implemented as an
+abstraction or even a separate service on the backend side with no changes to
+most HTTP servers.
+
+Backends are authenticated by their Ed25519 key, and the hash of that public key
+becomes the first segment of the path of client requests. To prevent abuse, it
+is RECOMMENDED that bastions apply allowlists for the backends that they serve.
+
+Like regular HTTPS reverse proxies, a bastion is in the position of blocking,
+observing, and manipulating the backend’s traffic. Bastions SHOULD only be used
+for self-authenticating protocols that don't rely exclusively on TLS for
+authentication or confidentiality. Alternatively, backends MAY use bastions they
+trust as part of their broad hosting infrastructure.
+
+Bastions were designed to enable synchronous [transparency log witnesses][]
+which prefer not to accept connections directly from the Internet.
+
+## API
+
+At a high level, backends connect to the bastion authenticating with their
+Ed25519 key and start serving HTTP/2 on that connection. The bastion then
+proxies to that connection all requests that include the backend's public key
+hash in the URL path.
+
+### Backend to bastion connection
+
+A backend connects to the bastion’s backend endpoint with TLS 1.3, specifying
+the ALPN protocol `bastion/0`. Other TLS versions MUST be rejected, as only TLS
+1.3 hides client certificates from network observers. The backend MUST present as
+its client certificate a self-signed Ed25519 certificate containing the backend
+public key. The backend verifies the bastion’s TLS certificate chain as usual.
+The bastion checks the backend public key against an allowlist.
+
+After opening the connection, the backend starts serving HTTP/2 traffic on it as
+if it was a client-initiated connection. HTTP/2’s multiplexing allows serving
+multiple parallel requests on a single connection efficiently. None of the
+backend's HTTP APIs are modified, except for observing the `X-Forwarded-For`
+header.
+
+Appendix B presents an example Go adapter that turns a regular HTTP server into
+a bastion backend.
+
+### Client to bastion requests
+
+The bastion accepts HTTP requests at
+
+```
+https://<bastion host>/<key hash>/<path>
```

Comment:
I think encouraging a dedicated host is fine, also because bastions need to make host-wide configuration changes to accept backend connections. Hosting a bastion alongside a different site is not technically impossible but is awkward enough that I think we can ignore it.